File: DataStructures/Java/CircularLinkedList.java
class CircularLinkedList {
    static class Node {
        // CSOFF: VisibilityModifier
        int data;
        Node next;
        // CSON: VisibilityModifier

        Node(final int newData) {
            this.data = newData;
            this.next = null;
        }
    }

    private Node head = null;

    public final void insertAtLast(final int data) {
        Node newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            this.head.next = this.head;
            return;
        }
        Node temp = this.head;
        while (temp.next != this.head) {
            temp = temp.next;
        }
        temp.next = newNode;
        newNode.next = this.head;
    }

    public void insertAtFirst(final int data) {
        Node newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            this.head.next = this.head;
            return;
        }
        Node temp = this.head;
        while (temp.next != this.head) {
            temp = temp.next;
        }
        newNode.next = this.head;
        this.head = newNode;
        temp.next = this.head;
    }

    public void insertAfter(final int data, final int after) {
        Node newNode = new Node(data);
        Node temp = this.head;
        while (temp != null) {
            if (temp.data == after) {
                newNode.next = temp.next;
                temp.next = newNode;
                return;
            }
            temp = temp.next;
        }
    }

    public void delete(final int data) {
        if (this.head == null) {
            return;
        }
        if (this.head.data == data) {
            Node temp = this.head;
            while (temp.next != this.head) {
                temp = temp.next;
            }
            temp.next = this.head.next;
            this.head = this.head.next;
            return;
        }
        Node temp = this.head;
        while (temp.next != this.head) {
            if (temp.next.data == data) {
                temp.next = temp.next.next;
                return;
            }
            temp = temp.next;
        }
    }

    public void print() {
        Node temp = this.head;
        if (temp != null) {
            do {
                System.out.print(temp.data + " -> ");
                temp = temp.next;
            } while (temp != this.head);
        }
        System.out.println("null");
    }

    public static void main(final String[] args) {
        CircularLinkedList list = new CircularLinkedList();

        for (int i = 1; i <= 5; i++) {
            list.insertAtLast(i);
        }

        list.print();

        list.insertAtFirst(0);
        list.insertAfter(6, 5);

        list.print();
    }
}

File: DataStructures/C/circular_linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct Node
{
    int data;
    struct Node *next;
    struct Node *prev;
};

struct Node *head = NULL;

void insert_at_last(int data)
{
    struct Node *new_node = (struct Node *)malloc(sizeof(struct Node));

    if (head == NULL)
    {
        new_node->data = data;
        new_node->next = new_node;
        new_node->prev = new_node;
        head = new_node;
    }
    else
    {
        struct Node *temp = head;
        while (temp->next != head)
        {
            temp = temp->next;
        }
        new_node->data = data;
        new_node->next = head;
        new_node->prev = temp;
        temp->next = new_node;
        head->prev = new_node;
    }
}

void insert_at_first(int data)
{
    struct Node *new_node = (struct Node *)malloc(sizeof(struct Node));

    new_node->data = data;

    new_node->next = head;
    new_node->prev = head->prev;

    head->prev->next = new_node;
    head->prev = new_node;

    head = new_node;
}

void delete_node(int data)
{
    if (head == NULL)
    {
        printf("List is empty\n");
        return;
    }

    struct Node *temp = head;
    struct Node *prev = NULL;

    if (head->data == data)
    {
        head->prev->next = head->next;
        head->next->prev = head->prev;
        head = head->next;
        free(temp);
        return;
    }

    while (temp->next != head)
    {
        if (temp->data == data)
        {
            prev->next = temp->next;
            temp->next->prev = prev;
            free(temp);
            return;
        }
        prev = temp;
        temp = temp->next;
    }

    if (temp->data == data)
    {
        prev->next = head;
        head->prev = prev;
        free(temp);
        return;
    }

    printf("Element not found\n");
}

void display()
{
    if (head == NULL)
    {
        printf("List is empty\n");
        return;
    }

    struct Node *temp = head;
    do
    {
        printf("%d ", temp->data);
        temp = temp->next;
    } while (temp != head);
    printf("\n");
}

int main()
{
    insert_at_last(10);
    insert_at_last(20);
    insert_at_last(30);
    insert_at_last(40);
    insert_at_last(50);
    display();

    insert_at_first(5);
    display();

    delete_node(5);
    display();

    delete_node(30);
    display();

    delete_node(50);
    display();

    delete_node(100);
    display();

    return 0;
}